swift/Sources/SignalClient/Aes256GcmEncryption.swift

```diff
@@ -0,0 +1,71 @@
+//
+// Copyright 2020 Signal Messenger, LLC
+// SPDX-License-Identifier: AGPL-3.0-only
+//
+
+import SignalFfi
+import Foundation
+
+public class Aes256GcmEncryption: ClonableHandleOwner {
+    public init<KeyBytes, NonceBytes, AssociatedDataBytes>(
+        _ key: KeyBytes,
+        _ nonce: NonceBytes,
+        _ associated_data: AssociatedDataBytes) throws
+        where KeyBytes: ContiguousBytes,
+              NonceBytes: ContiguousBytes,
+              AssociatedDataBytes: ContiguousBytes {
+
+        let handle: OpaquePointer? = try key.withUnsafeBytes { keyBytes in
+            try nonce.withUnsafeBytes { nonceBytes in
+                try associated_data.withUnsafeBytes { adBytes in
+                    var result: OpaquePointer?
+                    let error = signal_aes256_gcm_encryption_new(&result,
+                                                                 keyBytes.baseAddress?.assumingMemoryBound(to: UInt8.self),
+                                                                 keyBytes.count,
+                                                                 nonceBytes.baseAddress?.assumingMemoryBound(to: UInt8.self),
+                                                                 nonceBytes.count,
+                                                                 adBytes.baseAddress?.assumingMemoryBound(to: UInt8.self),
+                                                                 adBytes.count)
+                    try checkError(error)
+                    return result
+                }
+            }
+        }
+        super.init(owned: handle!)
+    }
+
+    internal override class func destroyNativeHandle(_ handle: OpaquePointer) -> SignalFfiErrorRef? {
+        return signal_aes256_gcm_encryption_destroy(handle)
+    }
+
+    public func encrypt<MutableBytes: ContiguousMutableBytes>(_ bytes: inout MutableBytes) throws {
+        try bytes.withUnsafeMutableBytes {
+            let error = signal_aes256_gcm_encryption_update(nativeHandle,
+                                                            $0.baseAddress?.assumingMemoryBound(to: UInt8.self),
+                                                            $0.count,
+                                                            0,
+                                                            UInt32($0.count))
+            try checkError(error)
+        }
+    }
+
+    public func encrypt<MutableBytes: ContiguousMutableBytes>(
+        _ bytes: inout MutableBytes,
+        _ offset: UInt32,
+        _ length: UInt32) throws {
+        try bytes.withUnsafeMutableBytes {
+            let error = signal_aes256_gcm_encryption_update(nativeHandle,
```

Comment:
FWIW I don't love this API. It's weird to me that the underlying rust takes in a mutable copy of the plaintext data and then mutates it into the ciphertext

Reply:
The intent is to avoid copies, since you're going to throw away the plaintext after you're done. I think we can do a little better for Swift, though, thanks to pervasive copy-on-write: you could have an API that passes an argument in by value, moves it to a local var, mutates it, and returns it, and that shouldn't do any copies if it's the only reference to the plaintext. (And if it wasn't, you'd have to do a copy anyway for the `inout` version.)

The other thing Swift has is cheap slices, so I don't think we need the offset and length parameters in the public API as long as ArraySlice also conforms to the new protocol.